internal/transports/p2p/server.go

```diff
@@ -29,99 +43,271 @@ func NewServer(
+		if err := s.connectToAddr(addr, s.chainParams.DefaultPort); err != nil {
+			continue
+		}
+
+		peersCounter++
+	}
+
+	if peersCounter == 0 {
+		return errors.New("cannot connect to any peers from seed")
 	}
 
-	inbound := false
-	conn, err := net.Dial(firstPeerAddr.Network(), firstPeerAddr.String())
+	s.log.Info().Msgf("connected to %d peers", peersCounter)
+	go s.observeOutboundPeers()
+	return nil
+}
+
+func (s *server) listenInboundPeers() error {
+	ourAddr := net.JoinHostPort("", fmt.Sprintf("%d", s.chainParams.DefaultPort))
+	listener, err := net.Listen("tcp", ourAddr)
 	if err != nil {
-		s.log.Error().Msgf("error connecting to peer %s, reason: %v", firstPeerAddr.String(), err)
+		s.log.Error().Msgf("error creating listener, reason: %v", err)
 		return err
 	}
 
-	return s.connectPeer(conn, inbound)
+	s.listener = listener
+	go s.observeInboundPeers()
+	return nil
 }
 
-func (s *server) listenAndConnect() error {
-	s.log.Info().Msgf("listening for inbound connections on port %s", s.chainParams.DefaultPort)
+func (s *server) connectToAddr(addr net.IP, port uint16) error {
+	netAddr := &net.TCPAddr{
+		IP:   addr,
+		Port: int(port),
+	}
 
-	ourAddr := net.JoinHostPort("", s.chainParams.DefaultPort)
-	listener, err := net.Listen("tcp", ourAddr)
+	conn, err := net.Dial(netAddr.Network(), netAddr.String())
 	if err != nil {
-		s.log.Error().Msgf("error creating listener, reason: %v", err)
+		s.log.Error().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("error connecting with peer, reason: %v", err)
+
+		s.log.Info().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	inbound := true
-	conn, err := listener.Accept()
+	peer, err := s.connectPeer(conn, false)
 	if err != nil {
-		s.log.Error().Msgf("error accepting connection, reason: %v", err)
+		s.log.Info().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	return s.connectPeer(conn, inbound)
+	_ = s.outboundPeers.AddPeer(peer) // don't need to check error here
+	s.addresses.UpsertPeerAddr(peer)
+	return nil
 }
 
-func (s *server) connectPeer(conn net.Conn, inbound bool) error {
-	peer, err := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log)
+func (s *server) connectPeer(conn net.Conn, inbound bool) (*peer.Peer, error) {
+	peer := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log, s)
+
+	err := peer.Connect()
 	if err != nil {
-		return err
+		peer.Disconnect()
+		s.log.Error().Str("peer", peer.String()).
+			Bool("inbound", inbound).
+			Msgf("error connecting with peer, reason: %v", err)
+		return nil, err
 	}
 
-	s.peers = append(s.peers, peer)
+	peer.SendGetAddrInfo()
```

Comment:
I would put it as a part of `Connect()` function, after the write/read/ping handlers and make it private.